Support condition for transactional operations (#899)
scalar-labs/scalardb: docs/api-guide.md

        .build();
```

##### Put with a condition

> 

Reviewer: I think it might be helpful to add something like this here.

You can write arbitrary conditions (e.g., a bank account balance must be equal to or more than zero) that you require a transaction to meet before being committed by having some checking logic in the transaction. Alternatively, you can also write simple conditions in a mutation operation, such as Put and Delete.

(If this is added, I think we need to update the next sentence a bit.)

Author: Thank you for the suggestion,  
To avoid redundancy in the explanation, I will merge the `Put with a condition` and `Delete with a condition` paragraphs.

Author: Fixed in https://github.com/scalar-labs/scalardb/pull/899/commits/46441d23a7772be58cd2ad2bfb887b71767f5b31

transaction.delete(delete);
```

#### Put and Delete with a condition

> You can write arbitrary conditions (e.g., a bank account balance must be equal to or more than zero) that you
> require a transaction to meet before being committed by having some checking logic in the transaction.
> Alternatively, you can also write simple conditions in a mutation operation, such as Put and Delete.

Reviewer 2: ```suggestion
You can write arbitrary conditions (e.g., a bank account balance must be equal to or more than zero) that you require a transaction to meet before being committed by having logic that checks the conditions in the transaction.
Alternatively, you can write simple conditions in a mutation operation, such as Put and Delete.
```

> When a Put or Delete operation includes a condition, it is only executed if the specified condition is met.

Reviewer 2: ```suggestion
When a Put or Delete operation includes a condition, the operation is executed only if the specified condition is met.
```

If the condition fails to be satisfied when the operation is executed, an exception called `UnsatisfiedConditionException` is thrown.

##### Conditions for Put
You can specify a condition in a Put operation as follows:

```java
// Build a condition
MutationCondition condition =
    ConditionBuilder.putIf(ConditionBuilder.column("c4").isEqualToFloat(0.0F))
        .and(ConditionBuilder.column("c5").isEqualToDouble(0.0))
        .build();

Put put =
    Put.newBuilder()
        .namespace("ns")
        .table("tbl")
        .partitionKey(partitionKey)
        .clusteringKey(clusteringKey)
        .floatValue("c4", 1.23F)
        .doubleValue("c5", 4.56)
        .condition(condition) // condition
        .build();

// Execute the Put operation
transaction.put(put);
```

> Other than the `putIf` condition, you can specify the `putIfExists` and `putIfNotExists` conditions as follows:

Reviewer 2: Does this part mean "Instead of using the `putIf` condition"?

```suggestion
Instead of using the `putIf` condition, you can specify the `putIfExists` and `putIfNotExists` conditions as follows:
```

Or does this part mean "In addition to using the `putIf` condition"?

```suggestion
In addition to using the `putIf` condition, you can specify the `putIfExists` and `putIfNotExists` conditions as follows:
```

Reviewer 2: @Torch3333 In that case, yes, let's use my second suggestion "In addition ... ."

```java
// Build a putIfExists condition
MutationCondition putIfExistsCondition = ConditionBuilder.putIfExists();

// Build a putIfNotExists condition
MutationCondition putIfNotExistsCondition = ConditionBuilder.putIfNotExists();
```

##### Conditions for Delete

You can specify a condition in a Delete operation as follows:

```java
// Build a condition
MutationCondition condition =
    ConditionBuilder.deleteIf(ConditionBuilder.column("c4").isEqualToFloat(0.0F))
        .and(ConditionBuilder.column("c5").isEqualToDouble(0.0))
        .build();

Delete delete =
    Delete.newBuilder()
        .namespace("ns")
        .table("tbl")
        .partitionKey(partitionKey)
        .clusteringKey(clusteringKey)
        .condition(condition)  // condition
        .build();

// Execute the Delete operation
transaction.delete(delete);
```

> Other than the `deleteIf` condition, you can specify the `deleteIfExists` condition as follows:

Reviewer 2: Does this part mean "Instead of using the `deleteIf` condition"?

```suggestion
Instead of using the `deleteIf` condition, you can specify the `deleteIfExists` condition as follows:
```

Or does this part mean "In addition to using the `deleteIf` condition"?

```suggestion
In addition to using the `deleteIf` condition, you can specify the `deleteIfExists` condition as follows:
```

Author: Same thinking as for the Put operation here. Either a `deleteIf` or a `deleteIfExists` condition can be specified for a Delete operation so the `In addition to using...` suggestion is more appropriate, isn't it?

Reviewer 2: @Torch3333 In that case, yes, let's use my second suggestion "In addition to using ... " here as well.

#### Notes

> All the builders of the CRUD operations can specify consistency with the `consistency()` methods, but it's ignored, and the `LINEARIZABLE` consistency level is always used in transactions.

Reviewer 2: I separated this sentence into two for readability.

```suggestion
Although all the builders of the CRUD operations can specify consistency by using the `consistency()` methods, those methods are ignored. Instead, the `LINEARIZABLE` consistency level is always used in transactions.
```

        } catch (RollbackException ex) {
          // Rolling back the transaction failed. You can log it here
        }

>       } catch (CrudException | CommitUnsatisfiedConditionException | CommitException e) {

Reviewer 3: The handling of `CommitUnsatisfiedConditionException` is very dependent on the application use cases. Therefore, we should use a separate `catch` block for this exception, as follows:

```suggestion
      } catch (CommitUnsatisfiedConditionException e) {
        // If you catch CommitUnsatisfiedConditionException, ...
      } catch (CrudException | CommitException e) {
```

Author: Understood, I put the new `UnsatisfiedConditionException` in a separate catch block. Thank you.
Fixed in https://github.com/scalar-labs/scalardb/pull/899/commits/46441d23a7772be58cd2ad2bfb887b71767f5b31
